src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,148 @@
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1, out _) ||
+                !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? streamType))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+                IMethodSymbol method = invocation.TargetMethod;
+                DiagnosticDescriptor rule;
+
+                if (string.Equals(method.Name, "ReadAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamReadAsyncMemoryOverloadsRule;
+                }
+                else if (string.Equals(method.Name, "WriteAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamWriteAsyncMemoryOverloadsRule;
+                }
+                else
+                {
+                    return;
+                }
+
+                if (IsPredecessorAwaiting(invocation) && // Verify if the top invocation in this statement is an await
+                    IsDefinedBy(method, streamType) && // Verify if the current method's type is or inherits from Stream
+                    HasUndesiredArguments(method))
+                {
+                    context.ReportDiagnostic(invocation.CreateDiagnostic(rule));
+                }
+            }, OperationKind.Invocation);
+        }
+
+        private static bool IsPredecessorAwaiting(IInvocationOperation invocation)
+        {
+            IOperation current = invocation;
+            while (current.Kind == OperationKind.Invocation)
```

Comment:
What is the situation you're looking for here where there's an invocation to Stream.Read/WriteAsync inside of another invocation?  And what happens if Parent ends up being null... don't we null ref?  And if we are inside of another invocation, doesn't that still mean the return type of the Read/WriteAsync call won't match when we change it?

I think we should be looking for (and only for) `await stream.Read/WriteAsync(...)`.  It might be easier/better rather than registering for OperationKind.Invocation, to instead register for OperationKind.Await, and then have the first check be whether the value of the await is an IInvocationOperation, and proceed from there.  I expect that'll result in the analyzer code being simpler, plus it should be faster, as there's almost certainly going to be many fewer awaits in a program than there are invocations.

Reply:
I simplified the analyzer code to only handle cases where we are directly awaiting the `ReadAsync/WriteAsync` methods, excluding `ContinueWith` or `ConfigureAwait`. Please take a look whenever you get a chance.

Reply:
> _I think we should be looking for (and only for) await stream.Read/WriteAsync(...)._

I based my latest commit on this comment, which is why I'm not currently handling `ConfigureAwait`, and I'm only handling `ReadAsync` and `WriteAsync`. Did I understand this incorrectly? In the other comments I am getting the suggestion to also handle `ConfigureAwait`.

Reply:
`ConfigureAwait` is sorta/kinda just part of the await.  It's generally `await (awaitable).ConfigureAwait(false)`.

Reply:
Done.